File: cf-mkv-extractor/src/reader.js
// Range-aware reader over an HTTP URL. Issues Range requests on demand
// and caches returned chunks so adjacent reads don't re-fetch. Targets
// the typical Matroska extraction pattern: read first ~1 MB, jump to
// the Cues offset near EOF, then read small subtitle clusters.

const DEFAULT_USER_AGENT =
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
    '(KHTML, like Gecko) Chrome/120.0 Safari/537.36';

export class RangeReader {
    constructor(url) {
        this.url = url;
        this.userAgent = DEFAULT_USER_AGENT;
        /** Total file size if known (Content-Range total). */
        this.totalSize = null;
        /** Cached chunks: [{ start, bytes }] keyed by start, sorted. */
        this._chunks = [];
    }

    /**
     * Returns a Uint8Array containing bytes [start, start+length).
     * Reuses cached chunks where possible, otherwise issues a single
     * Range GET and caches the result.
     */
    async read(start, length) {
        if (length <= 0) return new Uint8Array(0);
        // Hit cached chunk wholly containing the range?
        for (const c of this._chunks) {
            const end = c.start + c.bytes.length;
            if (c.start <= start && end >= start + length) {
                return c.bytes.subarray(start - c.start, start - c.start + length);
            }
        }
        // Otherwise fetch fresh.
        const bytes = await this._fetch(start, length);
        this._chunks.push({ start, bytes });
        // Keep cache bounded. Tight — 128 MB total memory cap is
        // shared with framework overhead + the lingering bytes
        // CF's stream layer holds after we cancel an open-ended
        // Range fetch (the cues read fallback). 12 MB cache + one
        // in-flight 6 MB batch buffer + cues slab + framework
        // leaves real headroom for those stream remnants.
        let cached = 0;
        for (const c of this._chunks) cached += c.bytes.length;
        while (cached > 12 * 1024 * 1024 && this._chunks.length > 2) {
            cached -= this._chunks.shift().bytes.length;
        }
        return bytes;
    }

    async _fetch(start, length, opts) {
        opts = opts || {};
        // Open-ended ranges (`bytes=N-` with no upper bound) are
        // a workaround for RD edges that drop bounded ranges at
        // deep offsets. _readStreamCapped caps the actual read at
        // `length` regardless, so the bandwidth cost is the same as
        // a bounded range — RD just sends data through a different
        // code path on its side.
        const rangeHeader = opts.openEnded
            ? `bytes=${start}-`
            : `bytes=${start}-${start + length - 1}`;
        const headers = {
            'Range': rangeHeader,
            'User-Agent': this.userAgent,
            'Accept': '*/*',
        };
        const res = await fetch(this.url, { headers, redirect: 'follow' });
        if (!res.ok && res.status !== 206) {
            throw new Error(`upstream HTTP ${res.status} on range ${rangeHeader}`);
        }
        // Capture total size from Content-Range: "bytes start-end/total"
        if (this.totalSize === null) {
            const cr = res.headers.get('content-range') || '';
            const m = /\/(\d+)$/.exec(cr);
            if (m) this.totalSize = parseInt(m[1], 10);
            if (this.totalSize === null) {
                const cl = res.headers.get('content-length');
                if (cl && res.status === 200) this.totalSize = parseInt(cl, 10);
            }
        }
        // If the upstream returned 200 OK instead of 206, the Range
        // header was ignored — the body is the entire file starting
        // at byte 0. CF's arrayBuffer() then predicts the read would
        // overflow the worker's memory cap and emits
        // "Memory limit would be exceeded before EOF" instantly.
        // Bail with a structured error so the caller can fall back
        // instead of OOMing for the user. (RD's standard CDN does
        // honour Range; this guards against a few of their mirrors
        // that misbehave.)
        if (res.status === 200 && start > 0) {
            try { res.body.cancel && res.body.cancel(); } catch (_) {}
            throw new Error(`upstream returned 200 (Range ignored) for ${rangeHeader}; cannot fetch arbitrary offset`);
        }
        // Stream the body chunk-by-chunk with a HARD CAP at the
        // requested length. Even if the upstream Content-Length lies
        // or the server streams more bytes than asked, we stop
        // reading at `length` and cancel the stream — keeps peak
        // memory bounded to one batch regardless of upstream weirdness.
        return await this._readStreamCapped(res, length);
    }

    async _readStreamCapped(res, maxBytes) {
        const reader = res.body.getReader();
        const buf = new Uint8Array(maxBytes);
        let written = 0;
        try {
            while (written < maxBytes) {
                const { value, done } = await reader.read();
                if (done) break;
                const room = maxBytes - written;
                if (value.length <= room) {
                    buf.set(value, written);
                    written += value.length;
                } else {
                    buf.set(value.subarray(0, room), written);
                    written = maxBytes;
                    break;
                }
            }
        } finally {
            // Fire-and-forget cancel: awaiting the cancel can deadlock
            // when the upstream connection has already dropped (the
            // promise never resolves and CF surfaces it as
            // "Network connection lost" against the worker as a whole).
            // Synchronously kick it off so the upstream socket gets
            // released eventually, but don't block on it.
            try { reader.cancel(); } catch (_) {}
        }
        return written === maxBytes ? buf : buf.subarray(0, written);
    }

    /**
     * read() with a graduated retry strategy targeting RD's deep-
     * offset drop behaviour. Three attempts:
     *
     *   1. Bounded at the requested length. Fast path for small
     *      reads at shallow offsets — head, tracks, cues fits in
     *      one go on most files.
     *   2. Larger bounded Range (up to ~8 MB). RD's CDN seems to
     *      treat mid-sized ranges differently from small ones —
     *      256 KB / 512 KB / 1 MB drops are common at deep
     *      offsets, but 8 MB ranges often go through. Possibly
     *      because they look more like normal browser streaming
     *      reads. We cache the larger chunk so subsequent reads in
     *      the same region hit cache.
     *   3. Open-ended Range ("bytes=N-") as last resort. RD streams
     *      "rest of file"; CF's stream layer holds the buffered
     *      tail in memory until GC, which can blow the 128 MB cap
     *      on big files. Only used when bounded approaches both
     *      fail, because the memory cost is real.
     *
     * Previous attempt (smaller-and-smaller splits 1/2/4) didn't
     * help: real-world logs show 256 KB drops at the same offsets
     * 1 MB drops. The size axis isn't where the recovery lives;
     * the LARGER bounded attempt + open-ended fallback is.
     */
    async readCritical(start, length) {
        const isTransient = (msg) =>
            msg.includes('Network connection lost') ||
            msg.includes('fetch failed') ||
            msg.includes('Memory limit') ||
            msg.includes('upstream HTTP 5');

        // ── Attempt 1: bounded at requested length ───────────────
        try {
            return await this.read(start, length);
        } catch (e) {
            const msg = (e && e.message) || String(e);
            if (!isTransient(msg)) throw e;
            console.warn(`[readCritical] start=${start} length=${length}: bounded request failed (${msg})`);
        }

        // ── Attempt 2: larger bounded Range ──────────────────────
        if (this.totalSize && start < this.totalSize) {
            const LARGER_SIZE = 8 * 1024 * 1024;
            const largerLen = Math.min(LARGER_SIZE, this.totalSize - start);
            if (largerLen > length) {
                await new Promise(r => setTimeout(r, 350));
                try {
                    console.log(`[readCritical] start=${start} length=${length}: trying larger bounded range of ${largerLen}`);
                    const big = await this._fetch(start, largerLen);
                    this._chunks.push({ start, bytes: big });
                    console.log(`[readCritical] start=${start} length=${length}: larger bounded ok`);
                    return big.subarray(0, length);
                } catch (e) {
                    const msg = (e && e.message) || String(e);
                    if (!isTransient(msg)) throw e;
                    console.warn(`[readCritical] start=${start} length=${length}: larger bounded failed (${msg})`);
                }
            }
        }

        // ── Attempt 3: open-ended Range (memory-risky) ───────────
        await new Promise(r => setTimeout(r, 700));
        try {
            console.log(`[readCritical] start=${start} length=${length}: trying open-ended Range`);
            const bytes = await this._fetch(start, length, { openEnded: true });
            this._chunks.push({ start, bytes });
            console.log(`[readCritical] start=${start} length=${length}: open-ended Range ok`);
            return bytes;
        } catch (e) {
            console.warn(`[readCritical] start=${start} length=${length}: open-ended Range also failed (${(e && e.message) || e})`);
            throw e;
        }
    }

    /**
     * Drops any cached chunk that starts inside [start, end). Used
     * by the extractor to release a batch's bytes the moment its
     * clusters have been parsed — keeps peak memory well below
     * Cloudflare's 128 MB Worker cap during long extractions.
     */
    evict(start, end) {
        for (let i = this._chunks.length - 1; i >= 0; i--) {
            const c = this._chunks[i];
            if (c.start >= start && c.start < end) {
                this._chunks.splice(i, 1);
            }
        }
    }

    /**
     * HEADs the file (well, GET with bytes=0-0) just to populate
     * totalSize. Useful at startup so later "read from end" math works.
     */
    async probeSize() {
        if (this.totalSize !== null) return this.totalSize;
        await this.read(0, 1);
        return this.totalSize;
    }
}

File: cf-mkv-extractor-rs/src/reader.rs
//! Range-aware reader over an HTTP URL. Port of reader.js.
//!
//! Two important differences from the JS version:
//!
//! 1. `_readStreamCapped` here drains the response body stream
//!    chunk-by-chunk via `Response::stream()`, stopping when we
//!    have `length` bytes. The stream is dropped immediately
//!    after — Rust's RAII means CF's underlying ReadableStream
//!    is cancelled the moment we leave scope, releasing any
//!    buffered tail bytes. The JS version's `reader.cancel()`
//!    in `finally` doesn't always cancel synchronously, which
//!    is one of the suspects for the memory pressure we saw
//!    on the open-ended Range fallback.
//!
//! 2. Cache uses `Vec<CachedChunk>` with explicit eviction, same
//!    semantics as the JS RangeReader cache.

use crate::is_transient_err;
use futures_util::StreamExt;
use std::time::Duration;
use worker::*;

const USER_AGENT: &str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 \
    (KHTML, like Gecko) Chrome/120.0 Safari/537.36";

/// Resident-cache cap. Smaller than the JS version's 40 MB because
/// the open-ended Range fallback's buffered tail eats into the
/// 128 MB cap; we'd rather keep the cache tight than chance an OOM.
const CACHE_CAP: u64 = 12 * 1024 * 1024;

struct CachedChunk {
    start: u64,
    bytes: Vec<u8>,
}

pub struct RangeReader {
    url: String,
    pub total_size: Option<u64>,
    chunks: Vec<CachedChunk>,
}

impl RangeReader {
    pub fn new(url: String) -> Self {
        Self { url, total_size: None, chunks: Vec::new() }
    }

    /// Returns the bytes for [start, start+length). Cache-hit when
    /// any single cached chunk wholly contains the range; otherwise
    /// fires a bounded fetch and caches the result.
    pub async fn read(&mut self, start: u64, length: u64) -> Result<Vec<u8>> {
        if length == 0 {
            return Ok(Vec::new());
        }
        for c in &self.chunks {
            let end = c.start + c.bytes.len() as u64;
            if c.start <= start && end >= start + length {
                let offset = (start - c.start) as usize;
                let len = length as usize;
                return Ok(c.bytes[offset..offset + len].to_vec());
            }
        }
        let bytes = self.do_fetch(start, length, false).await?;
        self.chunks.push(CachedChunk { start, bytes: bytes.clone() });
        self.evict_cache();
        Ok(bytes)
    }

    /// Reads we can't proceed without (head / tracks / cues).
    /// Three attempts:
    ///   1. Bounded at requested length (fast path).
    ///   2. Larger bounded Range (up to ~8 MB). RD's CDN seems
    ///      to treat mid-sized ranges differently from small ones
    ///      at deep offsets.
    ///   3. Open-ended Range — last resort, costs memory.
    pub async fn read_critical(&mut self, start: u64, length: u64) -> Result<Vec<u8>> {
        // Attempt 1: requested size.
        match self.read(start, length).await {
            Ok(bytes) => return Ok(bytes),
            Err(e) => {
                let msg = e.to_string();
                if !is_transient_err(&msg) {
                    return Err(e);
                }
                console_warn!(
                    "[readCritical] start={} length={}: bounded failed ({})",
                    start, length, msg
                );
            }
        }

        // Attempt 2: larger bounded Range.
        if let Some(total) = self.total_size {
            const LARGER: u64 = 8 * 1024 * 1024;
            if start < total {
                let larger_len = LARGER.min(total - start);
                if larger_len > length {
                    Delay::from(Duration::from_millis(350)).await;
                    console_log!(
                        "[readCritical] start={} length={}: trying larger bounded range of {}",
                        start, length, larger_len
                    );
                    match self.do_fetch(start, larger_len, false).await {
                        Ok(big) => {
                            console_log!(
                                "[readCritical] start={} length={}: larger bounded ok",
                                start, length
                            );
                            let result = big[..length as usize].to_vec();
                            self.chunks.push(CachedChunk { start, bytes: big });
                            self.evict_cache();
                            return Ok(result);
                        }
                        Err(e) => {
                            let msg = e.to_string();
                            if !is_transient_err(&msg) {
                                return Err(e);
                            }
                            console_warn!(
                                "[readCritical] start={} length={}: larger bounded failed ({})",
                                start, length, msg
                            );
                        }
                    }
                }
            }
        }

        // Attempt 3: open-ended Range.
        Delay::from(Duration::from_millis(700)).await;
        console_log!(
            "[readCritical] start={} length={}: trying open-ended Range",
            start, length
        );
        match self.do_fetch(start, length, true).await {
            Ok(bytes) => {
                console_log!(
                    "[readCritical] start={} length={}: open-ended Range ok",
                    start, length
                );
                self.chunks.push(CachedChunk { start, bytes: bytes.clone() });
                self.evict_cache();
                Ok(bytes)
            }
            Err(e) => {
                console_warn!(
                    "[readCritical] start={} length={}: open-ended Range also failed ({})",
                    start, length, e
                );
                Err(e)
            }
        }
    }

    /// Drops cached chunks whose start falls inside [start, end).
    /// Caller uses this to release a batch's bytes after parsing.
    pub fn evict(&mut self, start: u64, end: u64) {
        self.chunks.retain(|c| !(c.start >= start && c.start < end));
    }

    /// Ensures totalSize is known. Reads 1 byte at offset 0 if not.
    pub async fn probe_size(&mut self) -> Result<u64> {
        if let Some(t) = self.total_size {
            return Ok(t);
        }
        let _ = self.read(0, 1).await?;
        Ok(self.total_size.unwrap_or(0))
    }

    async fn do_fetch(&mut self, start: u64, length: u64, open_ended: bool) -> Result<Vec<u8>> {
        let range_header = if open_ended {
            format!("bytes={}-", start)
        } else {
            format!("bytes={}-{}", start, start + length - 1)
        };

        let mut headers = Headers::new();
        headers.set("Range", &range_header)?;
        headers.set("User-Agent", USER_AGENT)?;
        headers.set("Accept", "*/*")?;

        let mut init = RequestInit::new();
        init.with_method(Method::Get);
        init.with_headers(headers);
        init.with_redirect(RequestRedirect::Follow);

        let req = Request::new_with_init(&self.url, &init)?;
        let mut resp = Fetch::Request(req).send().await?;
        let status = resp.status_code();
        if status != 200 && status != 206 {
            return Err(Error::from(format!(
                "upstream HTTP {} on range {}",
                status, range_header
            )));
        }

        // Capture totalSize on first response that carries it.
        if self.total_size.is_none() {
            if let Ok(Some(cr)) = resp.headers().get("content-range") {
                if let Some(idx) = cr.rfind('/') {
                    if let Ok(n) = cr[idx + 1..].parse::<u64>() {
                        self.total_size = Some(n);
                    }
                }
            }
            if self.total_size.is_none() && status == 200 {
                if let Ok(Some(cl)) = resp.headers().get("content-length") {
                    if let Ok(n) = cl.parse::<u64>() {
                        self.total_size = Some(n);
                    }
                }
            }
        }

        // Range ignored — upstream returned 200 with full body.
        // Bail before we OOM on a multi-GB read.
        if status == 200 && start > 0 {
            return Err(Error::from(format!(
                "upstream returned 200 (Range ignored) for {}; cannot fetch arbitrary offset",
                range_header
            )));
        }

        // Stream the body chunk-by-chunk, hard cap at `length`.
        // For bounded Ranges length matches Content-Length so the
        // stream exhausts naturally. For open-ended Ranges we stop
        // early and drop the stream — that cancels CF's underlying
        // ReadableStream and releases the buffered tail.
        let mut stream = resp.stream()?;
        let max_bytes = length as usize;
        let mut buf: Vec<u8> = Vec::with_capacity(max_bytes);
        while buf.len() < max_bytes {
            match stream.next().await {
                Some(Ok(chunk)) => {
                    let room = max_bytes - buf.len();
                    if chunk.len() <= room {
                        buf.extend_from_slice(&chunk);
                    } else {
                        buf.extend_from_slice(&chunk[..room]);
                        break;
                    }
                }
                Some(Err(e)) => {
                    return Err(Error::from(format!("stream read failed: {}", e)));
                }
                None => break,
            }
        }
        // `stream` drops here → CF cancels the underlying reader →
        // tail bytes get released, no lingering memory pressure.
        Ok(buf)
    }

    fn evict_cache(&mut self) {
        let mut total: u64 = self.chunks.iter().map(|c| c.bytes.len() as u64).sum();
        while total > CACHE_CAP && self.chunks.len() > 2 {
            let removed = self.chunks.remove(0);
            total -= removed.bytes.len() as u64;
        }
    }
}
